scripts/linkchecker.py

```diff
@@ -0,0 +1,395 @@
+                 in-page links in the future.
+    :param anchor: Anchor string from the content page. This is provided to
+                help handle cases where target is empty.
+    :param target: The link target string to check
+    :returns: A checking record (string) if errors found, or None if we can
+              find the target link.
+    """
+    # B01: bad protocol
+    if target.startswith("http://"):
+        return new_record("ERROR", "Bad protocol", target)
+
+    # full link
+    if target.startswith("https://"):
+        # B03: self link, should revise to relative path
+        if (target.startswith("https://k8s.io/website/docs") or
+                target.startswith("https://kubernetes.io/website/docs")):
+            return new_record("ERROR", "Should use relative paths", target)
+        # external link, skip
+        return new_record("INFO", "External link", target)
+
+    # in-page link
+    # TODO: check if the target anchor does exists
+    if target.startswith("#"):
+        return new_record("INFO", "In-page link, skipped", target)
+
+    # Link has shortcode
+    if target.find("{{") > 0:
+        return new_record("WARNING", "Link has shortcode, skipped", target)
+
+    # link to English or localized page
+    if (target.startswith("/docs/") or
+            target.startswith("/" + ARGS.lang + "/docs/")):
+        # it is an embedded image
+        # TODO: an image might get translated as well
+        if target.endswith(".png"):
+            return new_record("INFO", "Link to image, skipped", target)
+
+        # target is shared reference (kubectl or kubernetes-api?
+        if (target.find("/docs/reference/generated/kubectl/") >= 0 or
+                target.find("/docs/reference/generate/kubernetes-api/") >= 0):
+            if check_file_exists(ROOT + "/static", target, "html"):
+                return None
+            return new_record("ERROR", "Missing shared reference", target)
+
+        # target is a markdown (.md) or a "<dir>/_index.md"?
+        if target.startswith("/docs/"):
+            base = os.path.join(ROOT, "content", "en")
+        else:
+            # localized target
+            base = os.path.join(ROOT, "content")
+        ok = check_file_exists(base, target)
+        if ok:
+            # We do't do additional checks for English site even if it has
+            # links to a non-English page
+            if ARGS.lang == "en":
+                return None
+
+            # If we are already checking localized link, fine
+            if target.startswith("/" + ARGS.lang + "/docs/"):
+                return None
+
+            # additional check for localization even if English target exists
+            base = os.path.join(ROOT, "content", ARGS.lang)
+            found = check_file_exists(base, target)
+            if not found:
+                # Still to be translated
+                return None
+            return new_record("ERROR", "Link not using localized page", target)
```

Comment:
I would think this is a useful check -- identifying if an `en` page exists and the localized page exists.
In reporting, you may want to emphasize the result or provide more information about the links.
What happens if the `en` link is invalid but a localized page exists?
Still looking: will the script identify if the link is in the _redirect file (yes) and what the link _should_ refer to (forwarded to)? I think you could save the data and add as a map or array of pairs?

Reply:
> What happens if the en link is invalid but a localized page exists?
If the link target is '/docs/...', and the English target is invalid, it is an error anyway. Though the real reason could be that the localization is out of sync with English.

> ... what the link should refer to
Good suggestion. I found it a painful process to manually check redirect entries. Will improve the script to suggest the correct target.